saq/queue/postgres.py

```diff
@@ -0,0 +1,608 @@
+            self.dequeue_timer_task = asyncio.create_task(
+                self.dequeue_timer(self.poll_interval)
+            )
+        self._connected = True
+
+    def job_id(self, job_key: str) -> str:
+        return job_key
+
+    def serialize(self, job: Job) -> json.Jsonb:
+        return json.Jsonb(job.to_dict())
+
+    def deserialize(self, job_dict: dict[t.Any, t.Any]) -> Job | None:
+        if job_dict.pop("queue") != self.name:
+            raise ValueError(f"Job {job_dict} fetched by wrong queue: {self.name}")
+        return Job(**job_dict, queue=self)
+
+    async def disconnect(self) -> None:
+        await self.pool.putconn(self.connection)
+        await self.pool.close()
+        self.wait_for_job_task.cancel()
+        self.listen_for_enqueues_task.cancel()
+        self.dequeue_timer_task.cancel()
+
+    async def info(
+        self, jobs: bool = False, offset: int = 0, limit: int = 10
+    ) -> QueueInfo:
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            await cursor.execute(
+                SQL(
+                    """
+                SELECT worker_id, stats FROM {stats_table}
+                WHERE %(now)s <= ttl
+                """
+                ).format(stats_table=self.stats_table),
+                {"now": seconds(now())},
+            )
+            results = await cursor.fetchall()
+        workers: dict[str, dict[str, t.Any]] = dict(results)
+
+        queued = await self.count("queued")
+        active = await self.count("active")
+        incomplete = await self.count("incomplete")
+
+        if jobs:
+            async with self.pool.connection() as conn, conn.cursor() as cursor:
+                await cursor.execute(
+                    SQL(
+                        """
+                    SELECT job FROM {jobs_table}
+                    WHERE status IN ('new', 'deferred', 'queued', 'active')
+                    """
+                    ).format(jobs_table=self.jobs_table),
+                )
+                results = await cursor.fetchall()
+            deserialized_jobs = (self.deserialize(result[0]) for result in results)
+            jobs_info = [job.to_dict() for job in deserialized_jobs if job]
+        else:
+            jobs_info = []
+
+        return {
+            "workers": workers,
+            "name": self.name,
+            "queued": queued,
+            "active": active,
+            "scheduled": incomplete - queued - active,
+            "jobs": jobs_info,
+        }
+
+    async def stats(self, ttl: int = 60) -> QueueStats:
+        stats = self._get_stats()
+        async with self.pool.connection() as conn:
+            await conn.execute(
+                SQL(
+                    """
+                INSERT INTO {stats_table} (worker_id, stats, ttl)
+                VALUES (%(worker_id)s, %(stats)s, %(ttl)s)
+                ON CONFLICT (worker_id) DO UPDATE
+                SET stats = %(stats)s, ttl = %(ttl)s
```

Comment:
is the stats table ever cleaned up? what does the ttl do?

Reply:
TTL is used by info to filter out stale stats. I'll also add something in the sweeper to delete old stats.

Reply:
what does redis do?

Reply:
The Redis queue:

1. Saves the current worker's stat to key "saq:stats:{worker_id}" with an expiration of ttl.
2. Removes stale worker_ids in the sorted set "saq:stats".
3. Add the current worker's id to the sorted set "saq:stats".
4. Sets an expiration of ttl on the sorted set "saq:stats".

Since we can't automatically expire rows in our stats tables, I'm using ttl to filter out stale stats in info() and add a step in the sweeper to delete them.